google-cloud-spanner/src/main/java/com/google/cloud/spanner/Interval.java

```diff
@@ -0,0 +1,316 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spanner;
+
+import com.google.auto.value.AutoValue;
+import com.google.errorprone.annotations.Immutable;
+import java.io.Serializable;
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.jetbrains.annotations.NotNull;
+
+@AutoValue
+@Immutable
+public abstract class Interval implements Serializable {
+  public static final long MONTHS_PER_YEAR = 12;
+  public static final long DAYS_PER_MONTH = 30;
+  public static final long HOURS_PER_DAY = 24;
+  public static final long MINUTES_PER_HOUR = 60;
+  public static final long SECONDS_PER_MINUTE = 60;
+  public static final long SECONDS_PER_HOUR = MINUTES_PER_HOUR * SECONDS_PER_MINUTE;
+  public static final long MILLIS_PER_SECOND = 1000;
+  public static final long MICROS_PER_MILLI = 1000;
+  public static final long NANOS_PER_MICRO = 1000;
+  public static final long MICROS_PER_SECOND = MICROS_PER_MILLI * MILLIS_PER_SECOND;
+  public static final long MICROS_PER_MINUTE = SECONDS_PER_MINUTE * MICROS_PER_SECOND;
+  public static final long MICROS_PER_HOUR = SECONDS_PER_HOUR * MICROS_PER_SECOND;
+  public static final long MICROS_PER_DAY = HOURS_PER_DAY * MICROS_PER_HOUR;
+  public static final long MICROS_PER_MONTH = DAYS_PER_MONTH * MICROS_PER_DAY;
+  public static final BigInteger NANOS_PER_SECOND =
+      BigInteger.valueOf(MICROS_PER_SECOND * NANOS_PER_MICRO);
+  public static final BigInteger NANOS_PER_MINUTE =
+      BigInteger.valueOf(MICROS_PER_MINUTE * NANOS_PER_MICRO);
+  public static final BigInteger NANOS_PER_HOUR =
+      BigInteger.valueOf(MICROS_PER_HOUR * NANOS_PER_MICRO);
+
+  private static final Pattern INTERVAL_PATTERN =
+      Pattern.compile(
+          "^P(?!$)(-?\\d+Y)?(-?\\d+M)?(-?\\d+D)?(T(?=-?.?\\d)(-?\\d+H)?(-?\\d+M)?(-?((\\d+(\\.\\d{1,9})?)|(\\.\\d{1,9}))S)?)?$");
+
+  /** Returns the months component of the interval. */
+  public abstract int months();
+
+  /** Returns the days component of the interval. */
+  public abstract int days();
+
+  /** Returns the microseconds component of the interval. */
+  public abstract long micros();
+
+  /** Returns the nanoFractions component of the interval. */
+  public abstract short nanoFractions();
+
+  public static Builder builder() {
+    return new AutoValue_Interval.Builder();
+  }
+
+  /** Returns the nanoseconds component of the interval. */
+  public BigInteger nanos() {
+    return BigInteger.valueOf(micros())
+        .multiply(BigInteger.valueOf(NANOS_PER_MICRO))
+        .add(BigInteger.valueOf(nanoFractions()));
+  }
+
+  /** Returns the total microseconds represented by the interval. */
+  public long getAsMicros() {
+    return months() * MICROS_PER_MONTH + days() * MICROS_PER_DAY + micros();
```

Comment:
Is this guaranteed not to overflow for any value that is returned by Spanner? If not, could we either:
1. Document what the behavior is (e.g. it silently overflows and returns gibberish)
2. OR check whether it would overflow and then return an error?

(Note: Java just silently overflows/underflows if you do nothing)